Library/Sender.ts

```diff
@@ -150,38 +156,112 @@ class Sender {
+        // Get the directory listing
+        fs.readdir(directory, (err, files) => {
+            if (err) {
+                return callback(err, -1);
+            }
+
+            let error: NodeJS.ErrnoException = null;
+            let totalSize = 0;
+            let count = 0;
+
+            if (files.length === 0) {
+                callback(null, 0);
+                return;
+            }
+
+            // Query all file sizes
+            for (let i = 0; i < files.length; i++) {
+                fs.stat(path.join(directory, files[i]), (err, fileStats) => {
+                    count++;
+
+                    if (err) {
+                        error = err;
+                    } else {
+                        if (fileStats.isFile()) {
+                            totalSize += fileStats.size;
+                        }
+                    }
+
+                    if (count === files.length) {
+                        // Did we get an error?
+                        if (error) {
+                            callback(error, -1);
+                        } else {
+                            callback(error, totalSize);
+                        }
+                    }
+                });
             }
         });
     }
 
+    /**
+     * Computes the size (in bytes) of all files in a directory at the root level. Synchronously.
+     */
+    private _getShallowDirectorySizeSync(directory: string): number {
+        let files = fs.readdirSync(directory);
+        let totalSize = 0;
+        for (let i = 0; i < files.length; i++) {
+            totalSize += fs.statSync(path.join(directory, files[i])).size;
+        }
+        return totalSize;
+    }
+
     /**
      * Stores the payload as a json file on disk in the temp directory
      */
     private _storeToDisk(payload: any) {
-
-        //ensure directory is created
+        // tmpdir is /tmp for *nix and USERDIR/AppData/Local/Temp for Windows
         var directory = path.join(os.tmpdir(), Sender.TEMPDIR_PREFIX + this._config.instrumentationKey);
 
+        // This will create the dir if it does not exist
+        // Default permissions on *nix are directory listing from other users but no file creations
+        Logging.info(Sender.TAG, "Checking existance of data storage directory: " + directory);
         this._confirmDirExists(directory, (error) => {
             if (error) {
+                Logging.warn(Sender.TAG, "Error while checking/creating directory: " + (error && error.message));
                 this._onErrorHelper(error);
                 return;
             }
 
-            //create file - file name for now is the timestamp, a better approach would be a UUID but that
-            //would require an external dependency
-            var fileName = new Date().getTime() + ".ai.json";
-            var fileFullPath = path.join(directory, fileName);
+            this._getShallowDirectorySize(directory, (err, size) => {
+                if (err || size < 0) {
```

Comment:
If the AI endpoint is inaccessible - will there be any perf issues in case we concurrently queue up a lot of storage operations such as calculate size of all files and then save a file? Or due to batching in SDK such calls will be rare and only happen couple times a minute, so no hit on storage throughput is expected?

Reply:
Batching is indeed the rate limit on this. On top of that, because this call is fully async (not blocking the event loop) I don't expect perf impact even if the batch size is configured to be very low.